doc news: add 14.1.0 entry (#2039)
groonga/groonga: doc/source/news/14.md

# News - 14 series

<!-- prettier-ignore-start -->
(release-14-1-0)=
## Release 14.1.0 - 2024-10-29
<!-- prettier-ignore-end -->

### Improvements

> ###### [status] `status`の結果に`llama.cpp`を追加しました。

Reviewer: ヘッダーの階層が少しずれていそうです 🙏🏾 
```suggestion
#### [status] `status`の結果に`llama.cpp`を追加しました。
```

Author: ちょっと、`#`が4つのれべるだと文字が大きすぎて見にくいので意図的に小さくしています。

Reviewer 2: それはセクションタイトルが長すぎることを示唆しているので、セクションレベルを大きくしないで読みやすくなるセクションタイトルに変えてください。

Reviewer 2: 基本的に、見た目のために論理的なマークアップを変えるのは悪手です。

Reviewer 2: たとえば、セクションタイトルは「`status`の結果に`llama.cpp`を追加しました。」のように「文」にする必要はありません。文にせずに端的にセクションの内容を示してください。

たとえば、「[status]」で`status`関連の話題というのを示しているので「 `status`の結果に」は不要です。

「[status] `.["features"]["llama.cpp"]`を追加」とかでも伝わるはずです。

Reviewer 2: セクションタイトルを短くしてセクションレベルを上げておいたので、後で確認してみてください。

Author: なるほど。
確認しました。ありがとうございます。次からセクションタイトルを端的に短くしてみます。

以下のように、`status`の結果の`features`に新しく`llama.cpp`の有効/無効を表示するようになっています。
(`status`の表示は長いので説明に必要な部分のみ抜粋しています。)

```
status
[
  [
    0,
    0,
    0.0
  ],
  {
    (省略)
    "features": {
      (省略)
      "llama.cpp": true,
      (省略)
    },
    (省略)
  }
]
```

`llama.cpp`は次に紹介する`language_model_vectorize()`に使用しています。

###### 新しい関数 `language_model_vectorize()` を追加しました。(実験的)

検索対象のドキュメントをベクトルに変換できます。

> 検索対象のドキュメントをベクトルとして保持できると、字面ではなく意味ベースの検索ができるようになります。

Reviewer: 意味ベースの部分を、「セマンティック検索」みたいに置き換える良かったりしますか？
- 読んでいて意味は現状のほうがわかりやすいと思うのですが、キャッチーなワードのほうが良いのかなと思ったため

Author: なるほど。ただ、すみません。私はわかりやすいのを重視したいので今の表現のままとさせてください。

Reviewer 2: 世間では「セマンティックサーチ」で通っているので「セマンティックサーチ」の方が伝わるはずです。「セマンティックサーチ」をまだ知らない人には「意味ベースの検索」の方が伝わりやすいかもしれませんが、そういう人をケアするならは「全文検索」も「キーワードによる検索」とかにした方がよいです。

が、ドキュメント全体を通してそういう人をケアする方向よりは、「用語集」みたいなところでケアして、ほかのところでは一般的な言い回しを用いたほうがよいです。

Author: 承知です。

例えば、「牛乳」と「ミルク」は字面は全然違いますが、同じ意味を持つ言葉です。
従来でも同義語展開を使えば、「牛乳」と「ミルク」のような字面は異なるが意味が同じ言葉を検索できますが、同義語展開用のテーブルをメンテナンスしなければなりません。

意味ベースの検索ができるようになると、同義語展開用のテーブルを準備、メンテナンスなどの特別なことをしなくても、Groongaがいい感じに近い言葉を探してくれるようになります。

###### [column_create] 新しい引数 `--generator` を追加しました。

カラムを自動生成できます。

> 以下のように他のカラムの値からカラムの値を自動で生成できます。

Reviewer: 以下のようにの「以下」の部分の例が無さそうです。

この機能のメリットは以下の2点です。

1. わざわざユーザーがテキスト生成処理を管理する必要がありません。
2. 生成元のカラムさえあれば良いので、`load`するデータ量を減らすことができます。

### Fixes

> ###### [{doc}`/server/gqtp`] GQPTクライアントに返却する`INVALID_ARGUMENT`の`RC`の値が`-22`ではなく、`65514`になっていた問題を修正しました。

Reviewer: ```suggestion
#### [{doc}`/server/gqtp`] GQPTクライアントに返却する`INVALID_ARGUMENT`の`RC`の値が`-22`ではなく、`65514`になっていた問題を修正しました。
```

      {
        "name": "content",
        "type": "Text"
      },
      {
        "name": "created_at",
        "type": "Time"
      },
      {
        "name": "n_likes",
        "type": "UInt32"
      },
      {
        "name": "tag",
        "type": "ShortText"
      }
    ],
    "records": [
    ],
    "records": [
      [
        "The first post!",
        "Welcome! This is my first post!",
        1436313600.0,
        5,
        "Hello"
      ],
      [
        "Groonga",
        "I started to use Groonga. It's very fast!",
        1436317200.0,
        10,
        "Groonga"
      ],
      [
        "Mroonga",
        "I also started to use Mroonga. It's also very fast! Really fast!",
        1436320800.0,
        15,
        "Groonga"
      ],
      [
        "Good-bye Senna",
        "I migrated all Senna system!",
        1436400000.0,
        3,
        "Senna"
      ],
      [
        "Good-bye Tritonn",
        "I also migrated all Tritonn system!",
        1436403600.0,
        3,
        "Senna"
      ]
    ]
  },
  "header": {
    "return_code": 0,
    "start_time": 1729647532.743954,
    "elapsed_time": 0.001991033554077148
  }
}
```

デフォルトのcommand versionは1です。

> お使いのGroongaのcommand versionは`status`コマンドで確認できます。

Reviewer: 確認している例があるとぱっとわかりやすいかもと思いました。

# News - 14 series

<!-- prettier-ignore-start -->
(release-14-1-0)=
## Release 14.1.0 - 2024-10-29
<!-- prettier-ignore-end -->

### Improvements

###### [status] `status`の結果に`llama.cpp`を追加しました。

以下のように、`status`の結果の`features`に新しく`llama.cpp`の有効/無効を表示するようになっています。
(`status`の表示は長いので説明に必要な部分のみ抜粋しています。)

```
status
[
  [
    0,
    0.0,
    0.0
  ],
  {
    (省略)
    "features": {
      (省略)
      "llama.cpp": true,
      (省略)
    },
    (省略)
  }
]
```

`llama.cpp`は次に紹介する`language_model_vectorize()`に使用しています。

> ###### 新しい関数 `language_model_vectorize()` を追加しました。(実験的)

Reviewer: ```suggestion
###### Added new function `language_model_vectorize()` (Experimental)
```

まだ実験的な機能なので安定していません。
また、この機能はCPUとメモリーを多く使います。

この関数は、与えられたテキストの正規化されたエンべディング（ベクトル表現）を返します。
エンべディングとは、与えられたテキストの意味情報をベクトルとして表現することです。
意味をベクトルとして表現できることで、字面ではなく意味ベースの検索ができるようになります。

> 例えば、「牛乳」と「ミルク」は字面は全然違いますが、同じ意味を持つ言葉です。
> 従来でも同義語展開を使えば、「牛乳」と「ミルク」のような字面の異なる意味が同じ言葉を検索できますが、同義語展開用のテーブルをメンテナンスしなければなりません。

Reviewer: ```suggestion
For example, "牛乳" ("milk" in kanji) and "ミルク" ("milk" in katakana) look completely different but share the same meaning.
Traditionally, you could search for words with different spellings but the same meaning by using synonym expansion.
But this required maintaining a synonym table.
```

> 意味ベースの検索ができるようになると、同義語展開用のテーブルを準備、メンテナンスせずに意味の近い言葉を検索できるようになります。

Reviewer: ```suggestion
With semantic search enabled, you can find words with similar meanings without preparing or maintaining a synonym table.
```

> ###### [column_create] 新しい引数 `--generator` を追加しました。

Reviewer: ```suggestion
###### [column_create] Added a new `--generator` argument.
```

> カラムを自動生成できます。
> 他のカラムの値からカラムの値を自動で生成できます。
> この機能のメリットは以下の2点です。

Reviewer: ```suggestion
You can now automatically generate columns based on the values of other columns.
The benefits of this feature include:
```

> 1. わざわざユーザーがテキスト生成処理を管理する必要がありません。
> 2. 生成元のカラムさえあれば良いので、`load`するデータ量を減らすことができます。

Reviewer: ```suggestion
1. Users no longer need to manage text generation processes manually.
2. You can reduce the amount of data to load since only the source columns are necessary.
```

> 例えば、他のカラムの値を`language_model_vectorize()`でエンべディングして、その値を自動生成したカラムに格納するということもできます。
> これを利用すれば、ユーザーは検索対象のテキストを`load`するだけで、エンべディングされたデータも同時に格納できるようになります。

Reviewer: ```suggestion
For example, you can use `language_model_vectorize()` to embed the values of other columns and store them in automatically generated columns.
This allows users to `load` only the text to be searched while simultaneously storing the embedded data.
```

### Fixes

> ###### [{doc}`/server/gqtp`] GQPTクライアントに返却する`INVALID_ARGUMENT`の`RC`の値が`-22`ではなく、`65514`になっていた問題を修正しました。

Reviewer: ```suggestion
###### [{doc}`/server/gqtp`] Fixed an issue where the `RC` value of `INVALID_ARGUMENT` returned to the GQTP client was `65514` instead of `-22`.
```

> GQTPクライアントを利用していないユーザーには影響ありません。

Reviewer: ```suggestion
This fix does not affect users who are not using GQTP clients.
```

> ###### [{doc}`/reference/commands/logical_range_filter`] command version 3 の時の応答に、余分な`"records":[]`が出力される問題を修正しました。

Reviewer: ```suggestion
###### [{doc}`/reference/commands/logical_range_filter`] Fixed an issue where an extra `"records": []` was output in the response when using command version 3.
```

> この問題は、command version 3の時のみ発生します。デフォルトのcommand versionは1です。
> お使いのGroongaのcommand versionは`status`コマンドで確認できます。

Reviewer: ```suggestion
This issue happens only with command version 3. The default command version is 1.
You can check your Groonga's command version using the `status` command.
```

```
status
[
  [
    0,
    0.0,
    0.0
  ],
  {
    (省略)
    "command_version": 1,
    "default_command_version": 1,
    (省略)
    "features": {
      (省略)
    },
    "apache_arrow": {
      (省略)
    },
    (省略)
  }
]
```

> 以下のように、余分な`"records":[]`が出力されますが、command version 3を使ってない場合は発生しません。

Reviewer: ```suggestion
As a follow, an extra `"records": []` was output.
This does not happen if you are not using command version 3.
```

      },
      {
        "name": "content",
        "type": "Text"
      },
      {
        "name": "created_at",
        "type": "Time"
      },
      {
        "name": "n_likes",
        "type": "UInt32"
      },
      {
        "name": "tag",
        "type": "ShortText"
      }
    ],
    "records": [
    ],
    "records": [
      [
        "The first post!",
        "Welcome! This is my first post!",
        1436313600.0,
        5,
        "Hello"
      ],
      [
        "Groonga",
        "I started to use Groonga. It's very fast!",
        1436317200.0,
        10,
        "Groonga"
      ],
      [
        "Mroonga",
        "I also started to use Mroonga. It's also very fast! Really fast!",
        1436320800.0,
        15,
        "Groonga"
      ],
      [
        "Good-bye Senna",
        "I migrated all Senna system!",
        1436400000.0,
        3,
        "Senna"
      ],
      [
        "Good-bye Tritonn",
        "I also migrated all Tritonn system!",
        1436403600.0,
        3,
        "Senna"
      ]
    ]
  },
  "header": {
    "return_code": 0,
    "start_time": 1729647532.743954,
    "elapsed_time": 0.001991033554077148
  }
}
```

> ###### [{doc}`/reference/commands/logical_range_filter`] {doc}`/reference/commands/request_cancel` でキャンセルした時の応答が、不正なJSONになる問題を修正しました。

Reviewer: ```suggestion
###### [{doc}`/reference/commands/logical_range_filter`] Fixed an issue where the response became invalid JSON when a request was canceled using {doc}`/reference/commands/request_cancel`.
```

> キャンセルのタイミングによって応答の出力が不正なJSONになる不具合を修正しました。

Reviewer: ```suggestion
A bug was fixed where the response output could become invalid JSON depending on the timing of the cancellation.
```

> ###### [{doc}`/reference/commands/index_column_diff`] 1つのレコードに同じトークンが大量に存在する時にインデックスの破損を誤検知する問題を修正しました。

Reviewer: ```suggestion
###### [{doc}`/reference/commands/index_column_diff`] Fixed an issue where index corruption was falsely detected when a single record contained a large number of identical tokens.
```

> この問題の発生条件は、一つのレコードに同一のトークンが13万個以上含まれていることですが、トークン数を簡単に確認する手段が無いので、
> `index_column_diff`を使って定期的にインデックスの破損を検出している場合は、念の為バージョンアップしておくことをおすすめします。

Reviewer: ```suggestion
This issue occurs when a single record contains more than 131,071 instances of the same token.
There's no easy way to check the number of tokens.
If you regularly use `index_column_diff` to detect index corruption, we recommend updating as a precaution.
```

# News - 14 series

<!-- prettier-ignore-start -->

> =======

Reviewer 2: ```suggestion
```
